File: src/utils/media.py
import mimetypes

from telegram import Bot, Message


async def get_sticker_image_bytes(
    message: Message, bot: Bot
) -> tuple[bytes, str] | None:
    sticker = message.sticker
    if not sticker:
        return None

    if sticker.is_animated or sticker.is_video:
        return None

    file = await bot.getFile(sticker.file_id)
    image_data = await file.download_as_bytearray()
    mime_type = (
        mimetypes.guess_type(file.file_path)[0] if file.file_path else None
    ) or "image/webp"
    return bytes(image_data), mime_type


async def get_message_image_bytes(
    message: Message,
    bot: Bot,
    *,
    allow_document: bool = False,
) -> tuple[bytes, str | None] | None:
    if message.photo:
        photo = message.photo[-1]
        file = await bot.getFile(photo.file_id)
        return bytes(await file.download_as_bytearray()), mimetypes.guess_type(
            file.file_path or ""
        )[0]
    if message.sticker:
        sticker_payload = await get_sticker_image_bytes(message, bot)
        if not sticker_payload:
            raise ValueError(
                "Animated/video stickers aren't supported yet. Send a static sticker or image."
            )
        return sticker_payload
    if (
        allow_document
        and message.document
        and (message.document.mime_type or "").startswith("image/")
    ):
        file = await bot.getFile(message.document.file_id)
        return bytes(await file.download_as_bytearray()), (
            message.document.mime_type or mimetypes.guess_type(file.file_path or "")[0]
        )
    return None

File: src/app/media.ts
import {
  downloadFile,
  Effect,
  messageMedia,
  type Bot,
  type BotApiError,
  type Message,
} from "telly";

export interface ImageData {
  readonly bytes: Uint8Array;
  readonly mimeType: string;
}

export function messageImage(
  message: Message,
  allowDocument = false,
): Effect.Effect<ImageData | undefined, BotApiError, Bot> {
  const media = messageMedia(message);
  if (media?.type === "photo") {
    return downloadFile({ fileId: media.photo.fileId }).pipe(
      Effect.map((bytes) => ({ bytes, mimeType: "image/jpeg" })),
    );
  }
  if (media?.type === "sticker") {
    if (media.sticker.isAnimated || media.sticker.isVideo) {
      return Effect.die(new Error(
        "Animated/video stickers aren't supported yet. Send a static sticker or image.",
      ));
    }
    return downloadFile({ fileId: media.sticker.fileId }).pipe(
      Effect.map((bytes) => ({ bytes, mimeType: "image/webp" })),
    );
  }
  if (
    allowDocument &&
    media?.type === "document" &&
    media.document.mimeType?.startsWith("image/") === true
  ) {
    const mimeType = media.document.mimeType;
    return downloadFile({ fileId: media.document.fileId }).pipe(
      Effect.map((bytes) => ({ bytes, mimeType })),
    );
  }
  return Effect.succeed(undefined);
}

export function imageDataUrl(image: ImageData): string {
  return `data:${image.mimeType};base64,${Buffer.from(image.bytes).toString("base64")}`;
}
